### PR TITLE
feat: ide: Convert while let to loop

### LINKED ISSUE #10901: Feature request: Add while-let-to-loop assist
Currently, Rust Analyzer only supports the assist to transform a plain `while` loop to a `loop {}`.

The desired assist should transform the following code (where `$pat`, `$expr` and `$body` are placeholders):

```rust
fn main() {
    while let $pat = $expr {
        $body
    }
}
```

to the code below when the cursor is placed inside the snippet `while let`:

```rust
fn main() {
    loop {
        if let $pat = $expr {
            $body
        } else {
            break;
        }
    }
}
```

Actually, this is exactly the desugaring rustc is doing (try `rustc <file> -Zunpretty=hir`).

### PR DESCRIPTION
**Pull request by rainy-me**
close #10901

### CODE REVIEW

File: crates/ide_assists/src/handlers/convert_while_to_loop.rs [hunk 1]
```diff
@@ -39,17 +39,14 @@ use crate::{
...
-    // Don't handle while let
-    if cond.pat().is_some() {
-        return None;
+    let if_cond = match while_cond.pat() {
+        Some(_) => while_expr.condition()?,
+        None => make::condition(invert_boolean_expression(while_cond.expr()?), None),
     };
-
-    let cond_expr = cond.expr()?;
-
```
Comment (Veykril) [inline crates/ide_assists/src/handlers/convert_while_to_loop.rs:52]: We can defer the construction with `make::condition` here by keeping the `let cond_expr = cond.expr()?;` here, then doing the match on `while_cond.pat()` in the builder closure.

Comment (rainy-me) [reply crates/ide_assists/src/handlers/convert_while_to_loop.rs:52]: I was not sure that when `while_cond.pat()` is some, is it guaranteed that `cond.expr()?` is some?

Comment (rainy-me) [reply crates/ide_assists/src/handlers/convert_while_to_loop.rs]: changed in https://github.com/rust-analyzer/rust-analyzer/pull/10916/commits/79b4e67b4ffdbdb58ad5b02398fd174e929609f8

Comment (Veykril) [reply crates/ide_assists/src/handlers/convert_while_to_loop.rs]: It is not guruanteed, but correct code requires both to be some(in fact correct code always has `expr` set), so we can safely make the assumption here as not supporting `while let Pat {}` shouldn't be a problem given it is not working code anyways.

Comment (rainy-me) [reply crates/ide_assists/src/handlers/convert_while_to_loop.rs]: Thanks for the explanation!

**Pull request merged**